src/library/dao/trackdao.cpp

```diff
@@ -630,49 +644,81 @@ bool TrackDAO::addTracksAdd(TrackInfoObject* pTrack, bool unremove) {
                     << "Failed to INSERT new track into library:"
                     << pTrack->getLocation();
-            return false;
+            return TrackId();
         }
         trackId = TrackId(m_pQueryLibraryInsert->lastInsertId());
+        m_analysisDao.saveTrackAnalyses(pTrack.data());
+        m_cueDao.saveTrackCues(trackId, pTrack.data());
+        DEBUG_ASSERT(!m_tracksAddedSet.contains(trackId));
+        m_tracksAddedSet.insert(trackId);
+    }
+    if (trackId.isValid()) {
         pTrack->setId(trackId);
-        m_analysisDao.saveTrackAnalyses(pTrack);
-        m_cueDao.saveTrackCues(trackId, pTrack);
+    } else {
+        qWarning() << "TrackDAO::addTracksAddTrack:"
+                << "Failed to add track to database"
+                << pTrack->getLocation();
+    }
+    // When exiting this function the trackId obtained from the
+    // database must match the id of the track, no matter if the
+    // track has been inserted or updated. If the trackId is still
+    // invalid adding the track to the library must have failed
+    // and the id of the track should also still be invalid.
+    DEBUG_ASSERT(pTrack->getId() == trackId);
+    // Only newly inserted tracks must be marked as clean!
+    // Existing or unremoved tracks have not been added to
+    // m_tracksAddedSet and will keep their dirty flag unchanged.
+    if (m_tracksAddedSet.contains(trackId)) {
         pTrack->markClean();
     }
-    m_tracksAddedSet.insert(trackId);
-    return true;
+    return trackId;
 }
 
-TrackId TrackDAO::addTrack(const QFileInfo& fileInfo, bool unremove) {
+TrackPointer TrackDAO::addTracksAddFile(const QFileInfo& fileInfo, bool unremove) {
+    // TODO(uklotzde): Resolve TrackInfoObject through TrackCache
     TrackPointer pTrack(new TrackInfoObject(fileInfo));
-    SoundSourceProxy(pTrack).loadTrackMetadata();
-    // Add the song to the database.
-    addTrack(pTrack.data(), unremove);
-    return pTrack->getId();
-}
 
-void TrackDAO::addTrack(TrackInfoObject* pTrack, bool unremove) {
-    //qDebug() << "TrackDAO::addTrack" << QThread::currentThread() << m_database.connectionName();
-    //qDebug() << "TrackCollection::addTrack(), inserting into DB";
-    // Why you be giving me NULL pTracks
-    DEBUG_ASSERT_AND_HANDLE(pTrack) {
-        return;
+    // Check that track is a supported extension.
+    if (!SoundSourceProxy::isFileSupported(fileInfo)) {
+        qWarning() << "TrackDAO::addTracksAddFile:"
+                << "Unsupported file type"
+                << pTrack->getLocation();
+        return TrackPointer();
     }
 
-    // Check that track is a supported extension.
-    if (!SoundSourceProxy::isFileSupported(pTrack->getFileInfo())) {
-        // TODO(XXX) provide some kind of error code on a per-track basis.
-        return;
+    // Initially load the metadata for the newly created track
+    // from the file.
+    SoundSourceProxy(pTrack).loadTrackMetadata();
```

Comment:
If the track is already in library, this call can be skipped. 
Fixing this requires some more refactoring. So please just add a TODO